torch/autograd/graph.py

```diff
@@ -629,3 +632,55 @@ def allow_mutation_on_saved_tensors():
         finally:
             ctx.clear()
             _allow_mutation_on_saved_tensors_enabled = False
+
+
+def _apply_hooks_on_graph_outputs(t_outputs: List[torch.Tensor]):
+    grad_fns = list(map(_get_grad_fn_or_grad_acc, t_outputs))
+
+    def iter_graph(roots):
+        if not roots:
+            return
+        seen = set()
+        q: Deque = collections.deque()
+        for node in roots:
+            if node is not None:
+                seen.add(node)
+                q.append(node)
+
+        while q:
+            node = q.popleft()
+            for fn, _idx in node.next_functions:
+                if fn in seen or fn is None:
+                    continue
+                seen.add(fn)
+                q.append(fn)
+
+            yield node
+
+    def prehook(grad_output):
+        node = torch._C._current_autograd_node()
+        log_str = f"Executing: {node} with grad_output: {grad_output}"
```

Comment:
nit: should we print the node unique identifier here?

Reply:
Hmm naively doing `id(node)` doesn't work - likely because the grad_fn PyObjects aren't being preserved and are continually being freed and being reallocated in the same place in memory as the backward is run.

Going to leave it as is for now for this version, but we can surface more info as follow up.